python/ray/tune/logger.py

```diff
@@ -119,6 +124,38 @@ def to_tf_values(result, path):
     return values
 
 
+class UtilMonitor(Thread):
+    def __init__(self, delay):
+        super(UtilMonitor, self).__init__()
+        self.stopped = False
+        self.delay = delay  # Time between calls to GPUtil
+        self.values = {}
+        self.start()
+
+    def read_utilization(self):
+        self.values.setdefault("perf/cpu", []).append(float(psutil.cpu_percent(interval=None)))
+        self.values.setdefault("perf/ram", []).append(float(getattr(psutil.virtual_memory(), 'percent')))
+        for gpu in GPUtil.getGPUs():
+            self.values.setdefault("perf/gpu"+str(gpu.id), []).append(float(gpu.load))
+            self.values.setdefault("perf/vram"+str(gpu.id), []).append(float(gpu.memoryUtil))
+
+    def get_data(self):
```

Comment:
I think you need a lock when entering get_data and read_utilization methods, to ensure thread-safety.

Reply:
threads in python are not really threads?

Reply:
https://stackoverflow.com/a/40072999